wire-runtime-swift/src/main/swift/ProtoCodable/ProtoWriter.swift

```diff
@@ -778,7 +778,7 @@ public final class ProtoWriter {
         let newCapacity = messageStackCapacity * 2
 
         let newMessageStack = UnsafeMutablePointer<MessageFrame>.allocate(capacity: newCapacity)
-        newMessageStack.moveInitialize(from: messageStack, count: messageStackIndex + 1)
+        newMessageStack.moveInitialize(from: messageStack, count: messageStackCapacity)
```

Comment:
Should this just be the stack index? (Eg just remove the +1)

Expanding the stack doesn't necessarily mean you're at the end of the stack, does it?

I think I'd want to see a few more test cases like:

Size is 4, write 2 bytes, then try to append a 4 byte message. I would guess that index is 2, and your change would shift the index to 4.

Reply:
i think we don't want to use `messageStackCapacity` since in theory it could still have uninitialized memory that we presumably should not read when copying it over into the new storage location.

IMO whatever the ultimate change, the code should probably be symmetric with the analogous logic in `ProtoReader` ([here](https://github.com/square/wire/blob/67b73b50ef79c063a2dacf675a15a9c65828371c/wire-runtime-swift/src/main/swift/ProtoCodable/ProtoReader.swift#L1001)), and that codepath should also probably get a comparable test case.

if the invariant we want is: '`messageStackIndex + 1` is always the number of initialized elements in the message stack buffer', then we should make sure that always holds. in this case maybe changing where the index increment & test happens would make the overall logic more straightforward & consistent. e.g. we'd change this: https://github.com/square/wire/blob/67b73b50ef79c063a2dacf675a15a9c65828371c/wire-runtime-swift/src/main/swift/ProtoCodable/ProtoWriter.swift#L711-L714

to something like what is done in `ProtoReader`: https://github.com/square/wire/blob/67b73b50ef79c063a2dacf675a15a9c65828371c/wire-runtime-swift/src/main/swift/ProtoCodable/ProtoReader.swift#L136-L140

Reply:
I had originally done it that way but was worried about some other issue cropping up. Since `ProtoReader` has similar code I'll go back to that version

Reply:
Typically if working if an unsafe language we wouldn't really allow this type of random access with pointer and size tracking a bit scattered. It would be great if we could do the same here. Maybe we could encapsulate the message stack into it's own class which closes up access to its internals, it might help a bunch with a few of the problems enumerated by @jamieQ.

Reply:
I agree, that was a much bigger refactor than I was prepared to do at the moment. I'd like to fix this issue without requiring a big refactor

Reply:
> I think I'd want to see a few more test cases like:
> 
> Size is 4, write 2 bytes, then try to append a 4 byte message. I would guess that index is 2, and your change would shift the index to 4.

This stack appears to be specific to length delimited fields, not bytes, and tracks the number of lengths that have been written or something along those lines. So in the test case it goes like this:

1. Start writing the entire message `Nested 1` with `try writer.encode(tag: 1, value: message)`
2. in `beginLengthDelimitedEncode`, increment `messageStackIndex` from -1 to 0.
3. Start writing the `nested` property of `Nested1` via `try protoWriter.encode(tag: 2, value: self.nested)` in `encode(to:)`
4. in `beginLengthDelimitedEncode`, increment `messageStackIndex` from 0 to 1.
5. Start writing the `nested` property of `Nested2` via `try protoWriter.encode(tag: 2, value: self.nested)` in `encode(to:)`
6. in `beginLengthDelimitedEncode`, increment `messageStackIndex` from 1 to 2.
7. Start writing the `nested` property of `Nested3` via `try protoWriter.encode(tag: 2, value: self.nested)` in `encode(to:)`
8. in `beginLengthDelimitedEncode`, increment `messageStackIndex` from 2 to 3.
9. Start writing the `nested` property of `Nested4` via `try protoWriter.encode(tag: 2, value: self.nested)` in `encode(to:)`
10. in `beginLengthDelimitedEncode`, increment `messageStackIndex` from 3 to 4.
11. Start writing the `nested` property of `Nested5` via `try protoWriter.encode(tag: 2, value: self.nested)` in `encode(to:)`
12. in `beginLengthDelimitedEncode`, increment `messageStackIndex` from 4 to 5.

At this point we are at index 5 which is equal to the initial capacity set here: 

https://github.com/square/wire/blob/67b73b50ef79c063a2dacf675a15a9c65828371c/wire-runtime-swift/src/main/swift/ProtoCodable/ProtoWriter.swift#L52

but we have not yet initialized that _index_. The number of bytes we have initialized is 5 (indices 0-4).

Now that `messageStackIndex == 5`, we hit the call to `expandMessageStack()` 
https://github.com/square/wire/blob/67b73b50ef79c063a2dacf675a15a9c65828371c/wire-runtime-swift/src/main/swift/ProtoCodable/ProtoWriter.swift#L712-L713

In `expandMessageStack()` it attempts to copy and deinitialize 6 bytes from the source buffer because `messageStackIndex + 1 == 6` 

https://github.com/square/wire/blob/67b73b50ef79c063a2dacf675a15a9c65828371c/wire-runtime-swift/src/main/swift/ProtoCodable/ProtoWriter.swift#L781

which copies garbage data into index 5. We then call `initialize(to:)` at index 5. This is potentially an invalid use of `initialize(to:)`, depending on whether the `MessageFrame` type is considered a "trivial type" as mentioned in the [docs](https://developer.apple.com/documentation/swift/unsafemutablepointer/initialize(to:)):

> The destination memory must be uninitialized or the pointer’s Pointee must be a trivial type